### PR TITLE
feat(be,fe): storage + smtp_request setup flow

### PR DESCRIPTION
**Pull request by sea-snake**
## Summary

PR 5+6 of the email-recovery stack — the storage layer plus the canister-side surface for the **setup (binding) flow**. Stacks on #3879 (DoH fallback).

The recovery half (`prepare_delegation`, `recover@id.ai` dispatch, delegation issuance) is deferred to a follow-up PR — it needs a stable reverse `address → AnchorNumber` index for the verified-`From:` lookup, which is best landed as its own focused change.

### What this PR adds

**Storage:**
- `EmailRecoveryCredential` lives **directly on the anchor struct** as `Option<EmailRecoveryCredential>` via a new `#[n(5)]` field on `StorableAnchor`. minicbor is forward-compatible across optional-field additions, so old anchors deserialize cleanly with `None` — no schema migration.
- Heap-only **pending-challenge map** (10k capacity, 30-min TTL, oldest-first eviction). Heap rather than stable: entries expire in 30 minutes and losing them on canister upgrade just costs an in-flight wizard one retry, not worth the stable-memory churn.
- **ChaCha20Rng PRNG**, seeded once per canister lifetime via `raw_rand`. Per-call cost is one PRNG draw, not a management-canister round trip.

**Canister surface (Candid):**
- `email_recovery_credential_prepare_add(IdentityNumber, EmailRecoveryDnsInput) -> Result<EmailRecoveryChallenge, EmailRecoveryError>` — authenticated update.
- `email_recovery_status(text) -> EmailRecoveryStatus` — anonymous query (FE polls).
- `email_recovery_credential_remove(IdentityNumber, text) -> Result<(), EmailRecoveryError>` — authenticated update.
- `smtp_request(SmtpRequest) -> SmtpResponse` — open update; the gateway entry point.

**Two verification paths (FE is path-agnostic):**

The FE just submits `{ address, selector, dns_proof: Option<DnsProofBundle> }`. The canister picks the path:

| Path | When | DKIM key sourced from | Outcall at email time? |
|---|---|---|---|
| **DNSSEC** | `dns_proof` is `Some` | Validated at prepare time → cached on pending challenge | No |
| **DoH allowlist** | `dns_proof` is `None` | `crate::doh::fetch_txt` (3-of-5 quorum, heap-cached) | Yes (first email per domain per TTL) |
| **Neither** | No proof + domain not allowlisted | — | Rejects with `DomainNotSupported` / `DomainNotAllowlisted` |

The DNSSEC path validates the chain synchronously against `DnssecConfig.root_anchors` and confirms the bundle's leaf is at `<selector>._domainkey.<domain>`. Candid-friendly mirror types (`DnsProofBundle`, `SignedRRset`, `Rrsig`, `DelegationLink`) are defined in the interface crate with `From<>` conversions to the canister-internal verifier types.

**Verification pipeline (smtp_request):**
- Recipient dispatch on the **full** address (`register@id.ai`) — defence-in-depth against direct callers spoofing just the user-part.
- Pulls the canister-issued nonce from the `Subject:` header (case-insensitive prefix match, hex-suffix length-checked).
- Sources the DKIM key (cached DNSSEC TXT or DoH fetch).
- Optionally fetches `_dmarc.<domain>`; absence forces strict alignment in the verifier.
- Runs `crate::dmarc::verify_email` end-to-end (full DKIM + DMARC + `Subject` in `h=`).
- On success: writes the credential to the anchor and flips the pending challenge to `Succeeded`. Failures flip to `Failed(reason)`. The gateway always sees `Ok`.

### What this PR does NOT add (intentional, deferred to follow-ups)

- **Recovery flow** — `prepare_delegation`, `recover@id.ai` dispatch, delegation issuance. Needs a reverse address index in stable storage; cleaner as its own PR.
- **Frontend** — the wizard UX from design-doc §8.6 is its own stack of PRs.

## Test plan

- [x] 441 unit tests pass: PRNG, pending-challenge map (TTL, eviction, status flips), prepare_add (input validation, allowlist gate, domain rejection, happy path), remove (happy path, case-insensitive match, idempotency edge cases), smtp.rs nonce extraction.
- [x] 12 PocketIC integration tests:
  - Nonce shape, allowlist gate, authz, status lifecycle (Pending → Expired), TTL expiry, silent-drop paths (no nonce, unknown nonce, wrong recipient).
  - **DNSSEC path**: no-trust-anchors rejection; DNSSEC-takes-precedence-over-DoH-allowlist.
  - **Full DoH e2e**: prepare → in-test DKIM signer → smtp_request → mocked DoH outcalls → RegistrationSucceeded → credential_remove succeeds.
- [x] `cargo check --target wasm32-unknown-unknown` clean.
- [x] `cargo clippy --tests -- -D warnings` clean.
- [x] `cargo fmt --check` clean.
- [x] `.did` interface stays in sync (the existing `check_candid_interface_compatibility` test enforces this).
- [x] Happy-path DNSSEC e2e test (with a real `_domainkey` chain).
- [ ] Smoke test on local replica — deferred to FE wizard integration.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## PR Stack
| # | PR | Description | Status |
|---|---|---|---|
| 0 | [#3836](https://github.com/dfinity/internet-identity/pull/3836) | Design doc | Open |
| 1 | [#3838](https://github.com/dfinity/internet-identity/pull/3838) | DNSSEC verifier scaffold | Open |
| 2 | [#3877](https://github.com/dfinity/internet-identity/pull/3877) | DKIM verifier (RFC 6376) | Open |
| 3 | [#3878](https://github.com/dfinity/internet-identity/pull/3878) | DMARC alignment (RFC 7489) | Open |
| 4 | [#3879](https://github.com/dfinity/internet-identity/pull/3879) | DoH fallback | Open |
| 5+6 | [#3880](https://github.com/dfinity/internet-identity/pull/3880) | Setup flow (storage + smtp_request) | Open |
| 7 | [#3881](https://github.com/dfinity/internet-identity/pull/3881) | Recovery flow (delegation) | Open |
| 8 | [#3882](https://github.com/dfinity/internet-identity/pull/3882) | Frontend + feature flag | Open |
| 9 | [#3883](https://github.com/dfinity/internet-identity/pull/3883) | Deploy/upgrade scripts: dnssec_config + doh_config | Open |
| 10 | [#3884](https://github.com/dfinity/internet-identity/pull/3884) | Email-recovery UX overhaul | Open |

## Changes during review

- The on-anchor `email_recovery` field is now a `Vec` (`Option<Vec<...>>` on the storable form, matching the existing `passkey_credentials`/`recovery_keys` pattern). The canister API still enforces ≤1 entry — this is a data-model widening so future multi-credential support doesn't need another schema bump. `Option<Vec<T>>` means anchors written under the previous schema decode cleanly with `email_recovery: None` (verified empirically against the in-tree minicbor-derive version).

### CODE REVIEW

**Review by timothyaterton — APPROVED** [review]
✅ **LGTM with follow-ups — ship it; see suggestions below.**

Thorough read of the storage layer, pending-challenge map, SMTP dispatch, both verification paths (DNSSEC + DoH), and the integration tests. The security model is solid and well-documented.

## Security

No blocking concerns. Specific strengths worth calling out:

- **No nonce oracle**: `smtp_request` always returns `Ok` regardless of the verification outcome — an open caller can't probe which nonces are in-flight.
- **Resource bounding**: 10K pending-map cap with oldest-first eviction, RFC 5321 address-size limits at every boundary (prepare, From:-extraction), and per-entry DKIM/DMARC TXT byte caps prevent heap inflation from crafted DNSSEC bundles or oversized addresses.
- **DNSSEC skeleton gating**: `verify_dnssec_skeleton` rejects bundles with >1 hop and pins the only allowed leaf to `_dmarc.<registered_domain>` — a DKIM leaf can't be smuggled in at prepare time.
- **`d=` zone anchoring** (`smtp.rs`): signatures whose `d=` doesn't anchor in the registered domain (or a strict subdomain) are rejected, preventing forwarder-re-signing bypass.
- **Recipient dispatch defence-in-depth**: `recipient_matches` checks both user-part and domain against `related_origins`, so a direct caller can't bypass dispatch by spoofing just the user-part with a different domain.
- **DoH DMARC degradation is safe**: `dmarc_bytes_opt = (..).ok()` means a DoH-fetch failure for the DMARC record falls back to "no DMARC published," which forces *strict* `d=` alignment in the verifier — the failure mode is more restrictive, not less.

## Maintainability

**[FOLLOW-UP]** `parse_txt_rdata` and `decode_dns_name_lowercase` are duplicated verbatim in `prepare.rs` and `submit_leaf.rs`. Worth extracting to a shared helper in the parent `email_recovery` module — duplicated parsing logic is a maintenance hazard if one copy gets a bugfix the other doesn't.

---

Ship it. **0 BLOCKING**, **1 FOLLOW-UP**.

**Review by aterga — APPROVED** [review]
**PR Review: #3880 (feat(email-recovery): storage + smtp_request setup flow)**

This is a solid, well-scoped incremental PR in a high-stakes feature (email recovery for Internet Identity anchors). It delivers the backend storage + setup binding flow exactly as described, with clean separation (email_recovery/ module, dedicated DKIM/DMARC/DNSSEC/DOH pieces), good test coverage (441 unit + 12 PocketIC integration), and clear defense-in-depth thinking (resource bounds, no-nonce-oracle design, DNSSEC primary + DoH allowlist fallback, strict DMARC degradation). CI is green, no merge conflicts, and the PR has already iterated on feedback (address caps, two-phase DNSSEC, multi-credential model).

No **major blockers** that prevent merging once the one outstanding comment is addressed. No obvious major security gaps. Below are the pragmatic, actionable items.

### 1. Blockers (must-fix before merge)
None critical. The only item that should be resolved is the **duplicated DNS parsing logic** already flagged by @timothyaterton:

- `parse_txt_rdata` and `decode_dns_name_lowercase` are duplicated between `prepare.rs` and `submit_leaf.rs`.
- **Action**: Extract them to a shared helper in `email_recovery/mod.rs` (or `dns.rs` if you prefer). One-line change, zero behavior change, reduces future maintenance risk. Do this + rebase/force-push.

Everything else (including the recent force-pushes for CI fixes and two-phase DNSSEC) looks ready.

### 2. Easy-to-change improvements (quick wins, low risk)
These are small, high-ROI tweaks you can land in the next push:

- **Add explicit security comments** on the three new public surfaces (`email_recovery_credential_prepare_add`, `smtp_request`, `email_recovery_credential_remove` and the status query). One-sentence invariant next to each Candid method impl in `main.rs` (e.g., "Caller must be the anchor controller – enforced by `check_caller_is_anchor_owner()`" and "smtp_request is a no-oracle endpoint: always returns Ok to avoid timing side-channels"). This makes future audits and onboarding trivial.
- **Rate-limit the new endpoints** (or at least document the current protection). The global 10 k pending-challenge map + 30 min TTL is a good backstop, but a simple per-anchor or per-caller counter (using the existing rate-limiter pattern already in II) would prevent one malicious anchor from starving the map. Trivial to add if not already present; otherwise just add a comment linking to the map capacity as the current mitigation.
- **Make the pending map eviction explicit and testable**. The 10 k cap + TTL is documented in the PR description — add a one-line `evict_expired()` call (or comment) in the `smtp_request` / status path so it's obvious that cleanup happens on every relevant call. Your unit tests already cover it; just surface it in the prod code for clarity.
- **Dependency hygiene** (Cargo.toml): The new crates for DKIM/DMARC/DNSSEC/DOH look correct, but pin any new ones to exact versions (or at least the same minor as the rest of the workspace) and add a comment explaining why each was added (e.g., "dnssec + doh for domain proof verification"). Prevents future supply-chain surprises.

All of the above are <30 min changes.

### 3. Security review (no major gaps)
The security model is thoughtful and matches the threat model for email-based recovery:

- **DNS verification paths**: DNSSEC (with proof + trust-anchor validation) as primary + DoH allowlist fallback (3-of-5 quorum, cached) is excellent. Two-phase flow (prepare → submit DKIM leaf) and selector matching are correct. DMARC failure → strict alignment (no leniency) is the right conservative choice.
- **Challenge / nonce handling**: ChaCha20Rng + heap-only pending map with hard bounds (10 k entries, 30 min TTL, address length caps) eliminates obvious DoS and oracle issues. The "smtp_request always returns Ok" design is explicitly called out — good.
- **Storage**: `EmailRecoveryCredential` lives on the anchor (stable memory, multi-credential support) — correct persistence model. Heap-only pending map is appropriate for short-lived challenges.
- **Authorization**: New methods follow II conventions (caller checks assumed present; confirm in the final diff).
- **Outcalls / cost**: DoH queries are the only new external surface; your PocketIC mocking and quorum logic already handle failures gracefully.

**No major security gaps found.** The biggest remaining risk is implementation subtlety in the new DKIM/DMARC/DNSSEC parsers (common source of bugs in email auth). Your test vectors and happy-path DNSSEC test mitigate this well.

**Recommended follow-up (non-blocking)**: After the full email-recovery stack lands, do a targeted security review (or formal audit) of the DKIM/DMARC verification path against the relevant RFCs (6376, 7208, 7489, etc.). Also verify RNG seeding comes from a strong source (IC randomness API or equivalent) — worth a one-line comment if not already obvious.

### 4. Strategic recommendations
- **Great feature priority**: Email recovery is one of the highest-UX improvements possible for II. Shipping the storage + setup flow now (frontend/recovery flow in later PRs) is the correct incremental approach.
- **Rate limiting & abuse monitoring**: Even with the 10 k cap, consider exposing a simple metric (pending map size) via a query or the existing II metrics endpoint. Long-term, you may want per-email-domain allowlisting or cooldowns if spam becomes an issue.
- **Cost awareness**: DNS outcalls (DoH) have cycle costs. Document the expected cost per `smtp_request` / prepare call so operators can set alerts. The quorum + caching design already keeps this reasonable.
- **Multi-credential & removal**: Supporting multiple recovery emails is smart. Make sure the removal method also cleans the pending map atomically (your tests appear to cover this).
- **Frontend / user messaging (for later PRs)**: When the frontend lands, be explicit with users that email recovery is **not** as strong as WebAuthn/passkeys — it shifts some trust to the email provider + DNS. A small warning + "recommended only as secondary factor" note would be prudent.
- **Upgrade safety**: Anchor storage change is non-trivial. Your tests cover it, but a quick manual upgrade test on a snapshot with existing anchors (pre-email-recovery) is worth doing before the final merge of the whole stack.

### Overall recommendation
**Approve after the duplicated DNS parsing extraction** (and the two tiny comment/rate-limit nits if you want to be extra clean). This PR is high-quality, well-tested, and moves the email-recovery feature forward without introducing new risk. The rest of the stack can land on top cleanly.

Let me know if you want me to deep-dive any specific file or test case — happy to review a follow-up diff or the recovery flow PR when it's ready. Great work!

File: src/internet_identity/src/email_recovery/remove.rs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+    anchor.email_recovery.remove(position);
+
+    // TODO(email-recovery): once the archive `Operation` enum gains
+    // an `EmailRecoveryRemove` variant, switch to it. Using the
+    // metadata-replace placeholder here keeps the type-checker happy
+    // without inventing new variants on a separate PR's surface.
+    Ok(Operation::IdentityMetadataReplace {
+        metadata_keys: vec![],
+    })
```
[the remark this answers is not in the document]
Comment (sea-snake) [reply src/internet_identity/src/email_recovery/remove.rs:68]: This is fine, archive is up for removal.

File: src/internet_identity/src/email_recovery/submit_leaf.rs [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+                    });
+                    Err(e)
+                }
+            }
+        }
+        Err(e) => {
+            let cloned = e.clone();
+            pending::with_mut(&nonce, now_secs, |c| {
+                c.status = PendingStatus::Failed(cloned);
+            });
```
[the remark this answers is not in the document]
Comment (aterga) [reply src/internet_identity/src/email_recovery/submit_leaf.rs:86]: Fixed in 05aa4cde0: cleared `partial_verification = None` in both `Failed` transitions in `submit_dkim_leaf`, matching the `Succeeded` branch.

File: src/internet_identity/src/email_recovery/smtp.rs [hunk 3]
```diff
@@ -0,0 +1,869 @@
...
+
+    // Reject simple/* on the header side. See design §5.2 — the
+    // canister-side verifier always uses relaxed header canonicalisation;
+    // simple-header signers are too rare to support.
+    if sig.c_header != crate::dkim::HeaderCanon::Relaxed {
+        return Err(EmailRecoveryError::EmailVerificationFailed(
+            "header canonicalisation must be relaxed".into(),
+        ));
+    }
+
```
[the remark this answers is not in the document]
Comment (aterga) [reply src/internet_identity/src/email_recovery/smtp.rs]: Updated in 05aa4cde0: the comment now reflects what the code actually does — we use the first DKIM-Signature header and reject (rather than walking the rest) if it doesn't parse or its `d=` falls outside the registered domain. The `d=` anchor check below enforces originating-domain authentication, so single-signature handling is deliberate, not a TODO.

**Comment by aterga** [comment]
Thanks for the reviews, @timothyaterton and @aterga. Disposition of every suggestion:

### Landed in this PR

- **Deduplicate `parse_txt_rdata` / `decode_dns_name_lowercase`** — flagged by both reviewers as the only blocker. Done in [3956ddd25](https://github.com/dfinity/internet-identity/pull/3880/commits/3956ddd25): both helpers moved to a new `email_recovery::dns` submodule with `pub(super)` visibility. `prepare.rs` (DMARC leaf) and `submit_leaf.rs` (DKIM leaf) now share one implementation. Pure refactor, zero behaviour change.
- **Clear `partial_verification` on `Failed` transitions in `submit_dkim_leaf`** — landed in [05aa4cde0](https://github.com/dfinity/internet-identity/pull/3880/commits/05aa4cde0). Matches the `Succeeded` branch and prevents the ~500B cached signature blob from sitting on terminal entries until TTL eviction.
- **Comment on DKIM-Signature header selection** ([smtp.rs:344](src/internet_identity/src/email_recovery/smtp.rs)) — rewritten in [05aa4cde0](https://github.com/dfinity/internet-identity/pull/3880/commits/05aa4cde0). The comment now reflects the deliberate policy: use the first DKIM-Signature header; reject on parse / `d=`-mismatch rather than walking the rest, because the `d=` anchor check enforces originating-domain authentication and walking would soften that.
- **`related_origins` in integration-test fixtures** — fixed in [7a4e77ce2](https://github.com/dfinity/internet-identity/pull/3880/commits/7a4e77ce2). Resolves the post-rebase CI failure on `full_setup_flow_binds_credential_to_anchor`.

### Already addressed / N/A

- **Archive `Operation` variant for `email_recovery_credential_remove`** — @sea-snake confirmed inline: archive is up for removal.

### Tracked in follow-up PR [#3896](https://github.com/dfinity/internet-identity/pull/3896) (stacks on top of [#3884](https://github.com/dfinity/internet-identity/pull/3884))

Non-blocking, documentation-only polish. Kept out of this PR to preserve focus and reviewability:

- **`Security:` stanzas on each new public method** in `main.rs` (authz, rate-limit, no-oracle invariants stated inline next to the implementation).
- **Document the 10 k pending-map cap + TTL as the per-canister rate-limit** on `prepare_add` / `prepare_delegation`. There's no per-anchor counter today; @aterga's review proposed one as a possible later addition.
- **Surface the lazy-eviction behaviour of `pending::with_mut`** in `smtp_request`'s pending lookup site, so future readers don't have to step through `pending.rs` to know when stale entries are dropped.
- **`Cargo.toml` hygiene** — add a comment explaining why `rand_chacha` was added and why `default-features = false` (the latter being load-bearing on wasm32).

### Punted to future work (non-blocking, called out in @aterga's review)

- **Real per-anchor / per-caller rate-limiter** on `prepare_add` / `prepare_delegation`. The 10 k map cap + TTL is the current backstop; a finer-grained counter would be a defence-in-depth addition, not a fix.
- **Targeted security review of the DKIM/DMARC/DNSSEC parsers** against the relevant RFCs after the full email-recovery stack lands.
- **Verify RNG seeding strength** — `rng.rs` already uses `raw_rand` once per canister lifetime to seed ChaCha20Rng; worth a one-line audit comment in a future pass.
- **`pending_map_size` metric** exposed via `/metrics` or a query — for ops visibility once the feature is live.
- **Manual upgrade-snapshot test** with existing anchors before the whole stack merges to main.

🤖 Generated with [Claude Code](https://claude.com/claude-code)
